src/layers/batch_normalization.rs

```diff
@@ -0,0 +1,236 @@
+use std::cell::Cell;
+
+use ndarray::RemoveAxis;
+
+use crate::{prelude::*, Layer};
+
+pub struct BatchNorm {
+    pub name: String,
+    pub features: usize,
+    pub eps: f64,
+    pub momentum: f64,
+    weight: Tensor<Ix1>,
+    bias: Tensor<Ix1>,
+    running_mean: Cell<Tensor<Ix1>>,
+    running_var: Cell<Tensor<Ix1>>,
+}
+
+impl Default for BatchNorm {
+    fn default() -> Self {
+        BatchNorm {
+            name: "Batch Normalization".into(),
+            features: 0,
+            eps: 1e-5,
+            momentum: 0.1,
+            weight: Tensor::ones(0),
+            bias: Tensor::zeros(0),
+            running_mean: Cell::new(Tensor::ones(0)),
+            running_var: Cell::new(Tensor::ones(0)),
+        }
+    }
+}
+
+impl BatchNorm {
+    pub fn new(name: impl ToString, features: usize) -> Self {
+        Self {
+            name: name.to_string(),
+            features,
+            weight: Tensor::ones(features),
+            bias: Tensor::zeros(features),
+            running_mean: Cell::new(Tensor::zeros(features)),
+            running_var: Cell::new(Tensor::ones(features)),
+            ..Default::default()
+        }
+    }
+
+    fn update_running_stat<D: Dimension>(
+        &self,
+        celled_running_stat: &Cell<Tensor<Ix1>>,
+        batch_stat: &Tensor<D>,
+    ) {
+        let scaled_batch_stat = (batch_stat * self.momentum)
+            .into_shape(batch_stat.len())
+            .unwrap();
+
+        let running_stat = celled_running_stat.take();
+        let scaled_running_stat = &running_stat * (1. - self.momentum);
+
+        let new_running_mean = scaled_batch_stat + scaled_running_stat;
+        celled_running_stat.set(new_running_mean);
+    }
+
+    fn iterative_mean<D: Dimension + RemoveAxis>(
+        &self,
+        input: &Tensor<D>,
+        axes: &[usize],
+    ) -> Tensor<D>
+    where
+        D::Smaller: Dimension<Larger = D>,
+    {
+        axes.iter()
+            .fold(input.clone(), |meaned, axis| self.mean::<D>(&meaned, *axis))
+    }
+
+    fn mean<D: Dimension + RemoveAxis>(&self, input: &Tensor<D>, axis: usize) -> Tensor<D>
+    where
+        D::Smaller: Dimension<Larger = D>,
+    {
+        let mean = input.mean_axis(Axis(axis)).unwrap();
+        mean.insert_axis(Axis(axis))
+    }
+
+    fn iterative_var<D: Dimension + RemoveAxis>(
+        &self,
+        input: &Tensor<D>,
+        axes: &[usize],
+        corrected: bool,
+    ) -> Tensor<D>
+    where
+        D::Smaller: Dimension<Larger = D>,
+    {
+        let shape = input.shape();
+        let biased = axes.iter().map(|axis| shape[*axis] as f64).product::<f64>();
+
+        let mean = self.iterative_mean(input, axes);
+        let normalized = (input - mean).mapv(|v| v.powf(2.0));
+
+        let var = self.iterative_mean(&normalized, axes);
```

Comment:
Using `ndarray`’s `var_axis` would lead to NaN values if unbiased, and 0 if biased, both of which would be wrong. Pytorch allows taking the variance across multiple dimensions, hence letting `n` be the amount of elements in each dimension, which `ndarray` does not support. Therefore, I am using the mean and then handling the biased or unbiased factor in the `match` statement below